src/attr/_make.py

```diff
@@ -871,22 +883,31 @@ def __init__(self, name, default, validator, repr, cmp, hash, init,
         bound_setattr("convert", convert)
         bound_setattr("metadata", (metadata_proxy(metadata) if metadata
                                    else _empty_metadata_singleton))
+        bound_setattr("type", type)
 
     def __setattr__(self, name, value):
         raise FrozenInstanceError()
 
     @classmethod
-    def from_counting_attr(cls, name, ca):
+    def from_counting_attr(cls, name, ca, type=None):
+        # type holds the annotated value. deal with conflicts:
+        if type is None:
+            type = ca.type
+        elif ca.type is not None:
+            raise ValueError(
+                "Type annotation and type argument are both present: "
```

Comment:
No need to add the type names into the message anymore.

Reply:
I thought it could be useful in more complex cases where it's not immediately obvious from the class definition how the type argument was set.  But I'm happy to drop it.